add conditional instantiation

ryandawsonuk opened the pull request:
for https://github.com/spring-cloud-incubator/spring-cloud-kubernetes/issues/183

pivotal-issuemaster commented:
@ryandawsonuk Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud-incubator/spring-cloud-kubernetes&pullRequestId=189)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud-incubator/spring-cloud-kubernetes&pullRequestId=189) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@ryandawsonuk Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud-incubator/spring-cloud-kubernetes&pullRequestId=189)!

spencergibb commented on spring-cloud-kubernetes-discovery/src/main/java/org/springframework/cloud/kubernetes/discovery/KubernetesDiscoveryClientAutoConfiguration.java:
```diff
@@ -11,6 +12,8 @@
 import org.springframework.context.annotation.Primary;
 
 @Configuration
+@EnableConfigurationProperties(KubernetesDiscoveryProperties.class)
```
No need for `(KubernetesDiscoveryProperties.class)` since it's declared as a bean below.

iocanel commented:
@ryandawsonuk can you please fix the test issues?

geoand commented:
Perhaps the failing test is flaky or there was some other issue. I wasn't able to reproduce the error locally. @iocanel would it be possible for you to relaunch the tests from the CircleCI dashboard?

aantono commented on spring-cloud-kubernetes-discovery/src/main/java/org/springframework/cloud/kubernetes/discovery/KubernetesDiscoveryClientAutoConfiguration.java:
```diff
@@ -11,6 +11,7 @@
 import org.springframework.context.annotation.Primary;
 
 @Configuration
+@ConditionalOnProperty(value = "spring.cloud.kubernetes.discovery.enabled", matchIfMissing = true)
```
Should also add `spring.cloud.kubernetes.enabled` condition as well, in case one decides to globally disable Spring Cloud Kubernetes altogether.

ryandawsonuk commented:
Updated PR following comments from @sta-szek, @aantono and @geoand

ryandawsonuk commented:
Sorry this now breaks the ribbon tests

ryandawsonuk commented:
Now the ribbon tests work again. It seems that adding ConditionalOnMissingBean(DiscoveryClient.class) to the bean instantiation for discoveryClient breaks them. I am not sure why but I don't think it's part of this PR so I've taken that out.

ryandawsonuk commented:
I've added tests for config (configmap and secrets) autoconfiguration to respect spring.cloud.kubernetes.enabled - it is already respecting it and I just added a test. For leader I've added the check and a test (the additional deps are test-scoped).

I notice there is a failure from circleci but mvn verify passes locally and the ci failure doesn't seem to be related.

iocanel commented:
@spencergibb: can you have a look please and approve if your concerns haven been taken care of?

ryandawsonuk commented:
@spencergibb let me know if there's anything you need from me on this

ArtyomGabeev commented:
@ryandawsonuk could you resolve merge conflicts?

@spencergibb any chance to get it merged? 

It will be nice to have both eureka/k8s spring cloud components in one code base enabled by properties/profiles/env.
